src/qibojit/backends/gpu.py

```diff
@@ -557,20 +527,32 @@ def calculate_eigenvectors(self, matrix, k: int = 6, hermitian: bool = True):
 
         return self.cp.linalg.eigh(matrix)
 
-    def calculate_matrix_exp(self, a, matrix, eigenvectors=None, eigenvalues=None):
+    def calculate_matrix_exp(
+        self,
+        matrix,
+        phase: Union[float, int, complex] = 1,
+        eigenvectors=None,
+        eigenvalues=None,
+    ):
         if eigenvectors is None or self.is_sparse(matrix):
-            if self.is_sparse(matrix):
-                from scipy.sparse.linalg import expm
+            is_sparse = self.is_sparse(matrix)
 
-                return self.cast(expm(-1j * a * self.to_numpy(matrix)))
+            if is_sparse:
+                from scipy.sparse.linalg import (  # pylint: disable=import-outside-toplevel
+                    expm,
+                )
             else:
                 from cupyx.scipy.linalg import expm  # pylint: disable=import-error
 
-                return self.cast(expm(-1j * a * matrix))
-        else:
-            expd = self.cp.diag(self.cp.exp(-1j * a * eigenvalues))
-            ud = self.cp.transpose(self.cp.conj(eigenvectors))
-            return self.cp.matmul(eigenvectors, self.cp.matmul(expd, ud))
+            _matrix = self.to_numpy(matrix) if is_sparse else matrix
+            _matrix = expm(phase * _matrix)
+
+            return self.cast(_matrix, dtype=_matrix.dtype)
```

Comment:
```suggestion
            if is_sparse:
                return self.cast(_matrix, dtype=_matrix.dtype)
            return _matrix
```
the cast is needed only the sparse case right? when you convert to numpy. I mean, it's also true that the cast does porbably nothing anyway even for cupy...

Reply:
Yeah the `cast` does nothing when the matrix is already a `cupy.ndarray`.